text/0000-export-visibility.md

```diff
@@ -0,0 +1,902 @@
+
+* Codegen tests that verify how `#[export_visibility = …]` is translated into
+  LLVM syntax
+* End-to-end tests for 3 platforms that cover ELF, Mach-O, and PE binaries.
+  Verification in such tests would most likely have to depend on arbitrary
+  developer tools (e.g.
+  [`readelf`](https://man7.org/linux/man-pages/man1/readelf.1.html) or
+  [`dumpbin`](https://learn.microsoft.com/en-us/cpp/build/reference/dumpbin-reference?view=msvc-170))
+  and therefore such tests would most likely have to be
+  `make`-based.
+
+## Support hidden visibility
+
+In the future, we may consider supporting `#[export_visibility = “hidden”]`.
+In terms of the internal `rustc` APIs this would map to
+[`rustc_target::spec::SymbolVisibility::Hidden`](https://github.com/rust-lang/rust/blob/910617d84d611e9ba508fd57a058c59b8a767697/compiler/rustc_target/src/spec/mod.rs#L884).
+The hidden visibility would have the following impact on Tier 1 binaries:
+
+* ELF binaries: The symbol is marked
+  [`STV_HIDDEN`](https://man7.org/linux/man-pages/man5/elf.5.html#:~:text=specific%20hidden%20class.-,STV_HIDDEN,-Symbol%20is%20unavailable)
+* PE binaries: The symbol is non-exported (i.e. the symbol is not listed in
+  [the `.edata` section](https://learn.microsoft.com/en-us/windows/win32/debug/pe-format#the-edata-section-image-only))
+* MachO binaries: The symbol is non-exported (i.e. the symbol is not listed in
+  [the export trie](https://github.com/apple-oss-distributions/xnu/blob/8d741a5de7ff4191bf97d57b9f54c2f6d4a15585/EXTERNAL_HEADERS/mach-o/loader.h#L1369))
+
+### Open question: `#[export_visibility = "hidden"]` vs `dylib`s
+[hidden-vs-dylibs]: #interaction-between-export_visibility--hidden-vs-dylibs
+
+#### Problem description
+
+https://github.com/rust-lang/rust/issues/73958#issuecomment-2635015556
+points out that using `#[export_visibility = "hidden"]` may break some `dylib`
+scenarios.
+
+For example, let's say that a crate named `rlib` is compiled into an `rlib` with
+the following functions:
+
+```rust
+/// Let's say that this is an internal helper that is only intended to be called
+/// from code within this library.  To facilitate this, this function is *not*
+/// `pub`.
+///
+/// To also enable calling the helper from a friendly (also internal-only),
+/// supporting C/C++ library we may use `#[no_mangle]`.  To keep this function
+/// internal and only enable directly calling this helper from statically-linked
+/// C/C++ libraries we may /// use `#[export_visibility = "hidden"]`.  We will
+/// see below how the hidden visibility may have some undesirable
+/// interactions with `dylib`s.
+///
+/// TODO: Do we need `rustc` command-line examples that would show how such
+/// static linking would be done when building a `staticlib`, `bin`, `cdylib`,
+/// or a `dylib` (I haven't checked how/if this would work in all of those
+/// cases;  i.e. I haven't checked how to ask `rustc` to link with static
+/// libraries produced by a C/C++ compiler).
```

Comment:
Let's please resolve this TODO before merging.  I'd suggest that, yes, it is probably helpful to describe exactly how to reproduce this.

Reply:
I think I'll just remove this TODO comment.

I wrote this comment long time ago, but I think that the "from [...] statically linked C/C++ libraries" was meant to
* justify using `#[no_mangle]` (because of "from [...] C/C++ libraries")
* justify using `#[export_visibility = "hidden"]` (because of "statically linked" and earlier "from a friendly [...] internal-only [...] C/C++ library")

But to understand and reproduce the problem described in this section, it is not necessary to actually do this (i.e. it is not necessary to statically link a C/C++ library).  It is sufficient to accept that `#[no_mangle]` and `#[export_visibility = "hidden"]` are justified.  And with that, it is possible to reproduce the problem using only Rust.

Reply:
I've removed the TODO in https://github.com/rust-lang/rfcs/commit/00584fec2b5565ca554aa7c2709394a09eb0897e